사다리 4-2, SK

sunghyun-k opened the pull request:
나중에 객체의 역할을 분담할때 다시 생각해보고 싶어 사다리의 구조체를 제거했습니다.
배열 확장에서 전에 가로대를 넣은 경우 다음엔 넣지 않도록 하기 위해 조건문을 사용해야 하는데 else문을 사용하지 말라는 요구사항이 있어 guard로 진행했으나 가독성이 더 떨어지는것 같아 if문으로 변경했습니다.

브랜치 생성과 사용에 있어서 꼬인것 같아 이전 브랜치를 완전히 지워버리고 코드를 넣게 되어서 커밋 내용이 완전히 사라진 것 같습니다... 우선 코멘트만 모아서 다시 입력 했습니다.

sunghyun-k committed:
사다리에 가로대를 삽입할 시 바로 전에 넣은 경우 넣지 않도록 수정
변수 및 함수명 일부 수정
사다리 구조체의 초기화 함수와 동일하다 생각되어 createLadder 함수 제거함.
사다리 구조체를 제거
else를 지우기 위해 guard를 사용해 보았으나 가독성이 떨어지는것으로 생각되어 사용하지 않음

godrm commented:
어딘가 브랜치에 커밋이 있으면 다른 브랜치에서도 얼마든지 가져올 수 있습니다. 
커밋 이력이 사라진건 아쉽습니다. 적정한 시점부터 새 브랜치를 만들고 다른 브랜치에서 커밋을 가져오는 걸 연습해보는 것을 추천합니다. PR을 보내지 않더라도 중요한 경험입니다.

sunghyun-k committed:
참가자 수 > 1, 사다리 높이 > 0 조건을 추가함.

sunghyun-k committed:
주석을 일부 수정함.

sunghyun-k committed:
사다리 배열 형태 변경
사다리 양 끝에 LadderComponent.empty를 넣어 나중에 사다리 게임을 실행할 시에 논리 구현에 용이하도록 수정함.

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -1,84 +1,83 @@
+            //                continue
+            //            }
+            //            self[index] = LadderComponent.rung
+            //            wasPlacedJustBefore = true
```
사용하지 않는 코드는 과감하게 지우고 커밋하세요. 커밋 어딘가에 남아있을 겁니다.

sunghyun-k replied:
넵. 지웠습니다.

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -1,84 +1,83 @@
+                wasPlacedJustBefore = true
+            } else {
+                wasPlacedJustBefore = false
             }
```
Array를 확장하는 시도는 좋습니다만,
mutating 함수를 만들기보다는 새로운 배열을 만들더라도 리턴하도록 순수한 함수를 만드는게 좋습니다.

sunghyun-k replied:
사다리 배열 입력에서 가로대를 추가한 사다리 배열을 반환하도록 재작성했습니다.

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -1,84 +1,83 @@
+    for index in ladder.indices {
+        ladder[index].insertRungsRandomlyWithoutSuccession()
+        ladder[index].insert(LadderComponent.empty, at: 0)
+        ladder[index].append(LadderComponent.empty)
```
맨앞, 맨뒤에는 왜 `.empty`를 넣는건가요?

sunghyun-k replied:
나중에 사다리 타기 로직을 구현할 때 용이하도록 추가하였습니다.

godrm replied:
아하 (실제로 사다리 타는 걸 구현하지는 않습니다 ㅎㅎ) 도전해보세요.

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -1,84 +1,83 @@
+    guard let height = Int(readLine()!), height > 0 else {
+        print("오류: 잘못된 형식")
         return nil
     }
```
여기서 두 번 입력받는 것도 일반화해서 함수를 분리해보세요.

sunghyun-k replied:
참여 인원과 높이를 입력받는 함수를 각각 나눴습니다.

godrm replied:
- 분리한 각각의 함수에서 오류 문구를 직접 출력하지 말고 예외 상황에 대해서 return 하거나 throw 하는 방법을 고민하세요.
- 두 함수의 공통 부분을 일반화해보세요.

sunghyun-k replied:
오류 처리 문법을 추가하고 입력함수 공통 부분을 함수로 묶었습니다.
오류 처리 문법을 처음 시도해봐서 이렇게 하는게 맞는지 모르겠네요. 개선할 점 피드백 부탁드립니다.

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -1,84 +1,83 @@
-    let ladder = getLadder(numberOfParticipants: input.numberOfParticipants, height: input.height)
-    print(ladder.stringizing())
+    let ladder = createLadder(numberOfParticipants: input.numberOfParticipants, height: input.height)
+    print(stringize(ladder: ladder))
```
run() 함수 입장에서 보면 하위에 getInputFromUser(), createLadder(), print() 함수가 있는 구조입니다.
앞에 두 개는 하위 함수가 맞는것 같은데, print() 구문은 직접 출력하는 코드가 run() 함수에 있어서 추상화 수준이 다르네요.
run() 함수가 print()를 하는지 안하는지 몰라도 되도록 출력함수를 분리해보세요.

sunghyun-k replied:
run 함수는 다른 하위 함수들을 실행하는 역할만 하도록 만들었습니다.

sunghyun-k replied:
커밋한 내용이 제대로 반영되지 않아 다시 한번 푸시했습니다.

godrm reviewed (commented):
다음 피드백을 확인해보시고, 함수를 분리해보세요.

sunghyun-k committed:
가로대 추가 함수 변경
자신을 변경하는 것이 아니고 다른 배열을 반환하도록 작성

sunghyun-k committed:
사용자 입력 함수 분할
사용자로부터 참여 인원과 사다리 높이를 입력 받는 함수를 나눔.

sunghyun-k committed:
run 출력 제거

sunghyun-k committed:
run 함수 수정
run은 다른 함수들을 실행하기만 하도록 변경함

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -1,84 +1,86 @@
-        }
-        return printableRow
+    return rowWithRung
+}
```
Array를 extension 해서 구현하는 방법 자체가 잘못된 건 아닙니다. 다른 스탭에서는 함께 고려해보세요.

sunghyun-k replied:
배열 확장이 더 마음에 들어 배열 확장으로 동일한 기능의 함수를 재작성했습니다.

sunghyun-k committed:
오류 처리 방식 변경

sunghyun-k committed:
오류를 처리하는 방식 수정
입력 함수들의 단위를 나누어 함수로 생성하여 코드를 줄임

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -51,36 +51,40 @@ func print(ladder: [[LadderComponent]]) {
-        print("오류: 잘못된 형식")
-        return nil
+    return try getCountableNumberFromUser()
+}
```
에러의 종류에 대해 고민해보세요.
최하위에 있는 `.notACountableNumber`가 의미있을까요? 
참여자 인원수에 대한 에러, 사다리 높이에 대한 에러 ... 이걸 구분할 수 있는게 좋을까요?
꼭 하위에 있는 에러를 최상위로 던져야 하는 건 아닙니다.

sunghyun-k replied:
여러 오류를 구분할 수 있도록 재작성했습니다.

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -51,36 +51,40 @@ func print(ladder: [[LadderComponent]]) {
-        return nil
-    }
-    return numberOfParticipants
+    return try getCountableNumberFromUser()
```
getNumberOfParticipants() 와 getLadderHeight() 에서 print() 부분은 일반화할 수 있을까요?

sunghyun-k replied:
출력 함수를 askUserAndGetNumber 함수에 내장했습니다.
이 함수는 질문 메시지를 문자열로 입력받습니다.

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -51,36 +51,40 @@ func print(ladder: [[LadderComponent]]) {
+    } catch InputError.notACountableNumber {
+        print("셀 수 있는 숫자가 아닙니다.")
+    } catch {
+        print("알 수 없는 오류 입니다.")
```
try - catch 를 사용할 때는 try가 여러 줄이 들어가기 보다는 do-try-catch 쌍이 각각 구분되는 게 좋습니다. 
(물론 더 좋은건 try-catch를 사용하지 않고 예외처리를 해서 에러를 출력하거나 흐름을 조정할 수 있는게 가장 좋습니다)

sunghyun-k replied:
오류 처리에 익숙하지 않아 일단은 스위프트 공식 문서를 참조하여 재작성 하였습니다.
catch 를 사용하지 않고 예외처리를 한다는 부분은 어떻게 진행하는지 모르겠습니다.
오류가 나지 않을 상황이면 최대한 try!를 사용하고,
try?를 사용해 nil일 경우 흐름을 제어하는 것을 말씀하시는 건가요?
오류 처리라는 것을 오늘 처음 배워서 아직까지 감이 잘 안잡힙니다 ^^;
많은 피드백 주시면 감사하겠습니다.

sunghyun-k committed:
배열 확장을 다시 넣음
배열의 메소드로 가로대를 넣도록 함. 기존 입력 후 반환하는 함수는 제거

sunghyun-k committed:
오류 처리 방식 수정

sunghyun-k committed:
질문을 출력하는 명령을 askUserAndGetNumber 함수에 내장함.

sunghyun-k committed:
오류 차리 방법 수정
